### PR TITLE
Configuration from environment variables.

### PR DESCRIPTION
**Pull request by brettmc**
Allow configuring resources, (some) exporters, samplers, and span processors from environment.
Add a TraceProviderFactory as a top-level method of creating a span processor from environment variables.
I think that other span exporters can be made to be env-configurable, but holding off until an upcoming improvement to use psr-7 discovery lands, which should simplify most of the exporters.

### CODE REVIEW

**Comment by brettmc** [comment]
This mostly fixes https://github.com/open-telemetry/opentelemetry-php/issues/302

**Comment by tidal** [comment]
Thank you.



> I think that other span exporters can be made to be env-configurable, but holding off until an upcoming improvement to use psr-7 discovery lands, which should simplify most of the exporters.

I am working on that atm. I hope it will simplify the exporters themself *, but there is no simple solution to it, since it's kind of a rabbit hole (there is not only the psr-7 issue). I will try to point out some specifics in a review.

______________
(*  or "themselves"? English is not my native language.)

File: src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php [hunk 1]
```diff
@@ -31,27 +32,41 @@ class BatchSpanProcessor implements SpanProcessorInterface
...
-        $this->maxExportBatchSize = $maxExportBatchSize;
+        $this->maxQueueSize = $maxQueueSize ?: $this->fromEnv('OTEL_BSP_MAX_QUEUE_SIZE', 2048);
+        $this->scheduledDelayMillis = $scheduledDelayMillis ?: $this->fromEnv('OTEL_BSP_SCHEDULE_DELAY', 5000);
+        $this->exporterTimeoutMillis = $exporterTimeoutMillis ?: $this->fromEnv('OTEL_BSP_EXPORT_TIMEOUT', 30000);
+        $this->maxExportBatchSize = $maxExportBatchSize ?: $this->fromEnv('OTEL_BSP_MAX_EXPORT_BATCH_SIZE', 512);
+        if ($this->maxExportBatchSize > $this->maxQueueSize) {
+            throw new InvalidArgumentException("maxExportBatchSize should be smaller or equal to $this->maxQueueSize");
+        }
 
         $this->queue = [];
```
Comment (tidal) [inline src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php:54]: I know, this has been there before, but could you please move this to the property declaration?

File: src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php [hunk 2]
```diff
@@ -31,27 +32,41 @@ class BatchSpanProcessor implements SpanProcessorInterface
...
+    {
+        $value = getenv($key);
+        if (false === $value) {
+            return $default;
+        }
+        if (!is_numeric($value)) {
+            throw new Exception($key . ' is not numeric');
+        }
+
+        return (int) $value;
```
Comment (tidal) [inline src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php:67]: This casting to int is redundant, since it's throwing a exeption before, when it's not.

File: src/SDK/Trace/TracerProvider.php [hunk 3]
```diff
@@ -33,7 +33,7 @@ public function __construct(
         }
 
         $spanProcessors = is_array($spanProcessors) ? $spanProcessors : [$spanProcessors];
-        $resource = $resource ?? ResourceInfo::defaultResource();
+        $resource = $resource ?? ResourceInfo::create(new Attributes());
```
Comment (tidal) [inline src/SDK/Trace/TracerProvider.php:36]: Why did you change this?

File: src/SDK/Trace/TracerProviderFactory.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        ?SamplerFactory $samplerFactory = null,
+        ?SpanProcessorFactory $spanProcessorFactory = null
+    ) {
+        $this->name = $name;
+        $this->exporterFactory = $exporterFactory ?: new ExporterFactory($this->name);
+        $this->samplerFactory = $samplerFactory ?: new SamplerFactory();
+        $this->spanProcessorFactory = $spanProcessorFactory ?: new SpanProcessorFactory();
+    }
+
+    public function create(): API\TracerProviderInterface
```
Comment (tidal) [inline src/SDK/Trace/TracerProviderFactory.php:28]: The semantics of the method suggest, that this is the only way to create a TracerProvider, while there could be possibly more. So either this class could be renamed appropriatly or the method. 
Type hinting `TracerProviderInterface` suggests, that there can be possibly more factories, and this factory is implementing an interface.

File: tests/SDK/Unit/Resource/ResourceTest.php [hunk 5]
```diff
@@ -110,4 +120,60 @@ public function testImmutableCreate()
...
+    /**
+     * @test
+     * @dataProvider environmentResourceProvider
+     */
+    public function resource_fromEnvironment(string $envAttributes, array $userAttributes, array $expected)
+    {
+        $this->setEnvironmentVariable('OTEL_RESOURCE_ATTRIBUTES', $envAttributes);
+        $resource = ResourceInfo::create(new Attributes($userAttributes));
+        foreach ($expected as $name => $value) {
+            $this->assertEquals($value, $resource->getAttributes()->get($name));
```
Comment (tidal) [inline tests/SDK/Unit/Resource/ResourceTest.php:133]: Minor issue, but 'assertEquals' is not type safe, while 'assertSame' is.

File: tests/SDK/Unit/Trace/SpanLimitsBuilderTest.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    use EnvironmentVariables;
+
+    /**
+     * @test
+     */
+    public function spanLimitsBuilder_usesDefaultValues()
+    {
+        $builder = new SpanLimitsBuilder();
+        $spanLimits = $builder->build();
+        $this->assertEquals(128, $spanLimits->getAttributeLimits()->getAttributeCountLimit());
```
Comment (tidal) [inline tests/SDK/Unit/Trace/SpanLimitsBuilderTest.php:23]: Where does this magic number come from?

**Comment by tidal** [comment]
The Phan errors in CI should be fixed, when https://github.com/open-telemetry/opentelemetry-php/pull/470 is merged.

File: composer.json [hunk 7]
```diff
@@ -51,6 +51,7 @@
         }
     },
     "require-dev": {
+        "assertwell/phpunit-global-state": "^0.2",
```
[the remark this answers is not in the document]
Comment (brettmc) [reply composer.json:54]: no, the package is just being used for the EnvironmentVariables trait - to set/clear environment variables via putenv (which phpunit does not do).

File: src/SDK/Resource/ResourceInfo.php [hunk 8]
```diff
@@ -69,10 +73,34 @@ public static function defaultResource(): self
...
     }
 
+    /**
+     * Create resource attributes from OTEL_RESOURCE_ATTRIBUTES key=value comma-separated entries
+     * @see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#specifying-resource-information-via-an-environment-variable
+     */
+    public static function environmentResource(): self
+    {
+        $attributes = [];
+        $string = getenv('OTEL_RESOURCE_ATTRIBUTES');
```
[the remark this answers is not in the document]
Comment (brettmc) [reply src/SDK/Resource/ResourceInfo.php:88]: A quick search only found c++ as a potentially unsafe implementation of `getenv`. I chose getenv over $_ENV for two reasons: it's already in use where we already retrieve env vars (in the otel exporters), and I believe it's advised not to access superglobals directly.
Do you think it would be worth creating a trait so that all env var handling is consistent? I already found myself duplicating a small function in a couple of places when handling env vars...

File: src/SDK/Resource/ResourceInfo.php [hunk 9]
```diff
@@ -69,10 +73,34 @@ public static function defaultResource(): self
...
+        if ($string && false !== strpos($string, '=')) {
+            foreach (explode(',', $string) as $pair) {
+                list($key, $value) = explode('=', $pair);
+                $attributes[trim($key)] = trim($value);
+            }
+        }
+        //@see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#general-sdk-configuration
+        $serviceName = getenv('OTEL_SERVICE_NAME');
+        if ($serviceName) {
```
[the remark this answers is not in the document]
Comment (brettmc) [reply src/SDK/Resource/ResourceInfo.php:97]: I agree that this could be a warning. I had started adding some logging in this branch for situations like this (for the eventual implementation of "must not throw exceptions"), but decided to keep this PR on-topic and not do it yet. It's amongst the next couple of things I want to work on :)

File: src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php (hunk 2 above)
Comment (brettmc) [reply src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php:67]: It's actually required, because getenv always returns a string, and if you don't cast it you get a TypeError. You could argue that the exception is not required, but I thought the meaningful exception message was better than a type error.

File: src/SDK/Trace/TracerProvider.php (hunk 3 above)
Comment (brettmc) [reply src/SDK/Trace/TracerProvider.php:36]: So that environment variables are picked up.

File: src/SDK/Resource/ResourceInfo.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (tidal) [reply src/SDK/Resource/ResourceInfo.php:88]: Yeah, sorry couldn't find it either. Maybe I mixed it up with putenv (which is only used in tests). What you say makes sense. :)

File: src/SDK/Resource/ResourceInfo.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (tidal) [reply src/SDK/Resource/ResourceInfo.php:97]: Ok great!

File: src/SDK/Trace/TracerProvider.php (hunk 3 above)
Comment (tidal) [reply src/SDK/Trace/TracerProvider.php:36]: Hmm, I don't get it, but I'll just believe you :)

File: examples/ConfigurationFromEnvironment.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+putenv('OTEL_TRACES_SAMPLER_ARG=0.95');
+putenv('OTEL_TRACES_EXPORTER=console');
+putenv('OTEL_TRACES_PROCESSOR=batch'); //TODO not in spec, @see https://github.com/open-telemetry/opentelemetry-specification/issues/2110
+putenv('OTEL_BSP_SCHEDULE_DELAY=10000');
+
+echo 'Creating Exporter From Environment' . PHP_EOL;
+
+$tracerProvider = (new TracerProviderFactory('example'))->create();
+
+$tracer = $tracerProvider->getTracer('io.opentelemetry.contrib.php');
```
[the remark this answers is not in the document]
Comment (brettmc) [reply examples/ConfigurationFromEnvironment.php:22]: I thought the same as you - as I understand it, it would be unusual for that _not_ to be the tracer name. I will make it a default.

File: examples/ConfigurationFromEnvironment.php [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+putenv('OTEL_LOG_LEVEL=warning');
+putenv('OTEL_TRACES_SAMPLER=traceidratio');
+putenv('OTEL_TRACES_SAMPLER_ARG=0.95');
+putenv('OTEL_TRACES_EXPORTER=console');
+putenv('OTEL_TRACES_PROCESSOR=batch'); //TODO not in spec, @see https://github.com/open-telemetry/opentelemetry-specification/issues/2110
+putenv('OTEL_BSP_SCHEDULE_DELAY=10000');
+
+echo 'Creating Exporter From Environment' . PHP_EOL;
+
+$tracerProvider = (new TracerProviderFactory('example'))->create();
```
[the remark this answers is not in the document]
Comment (brettmc) [reply examples/ConfigurationFromEnvironment.php:20]: Yes, definitely. I did have it catching those and setting null/noop as appropriate, but I thought that was too mysterious without having the capability to log errors and warnings. I will work on a follow-up PR to add logging, and then we can work towards not throwing any exceptions.
(Or, did you just mean the example would be better with a try/catch to illustrate that it can throw exceptions?)

File: src/SDK/Trace/ExporterFactory.php [hunk 12]
```diff
@@ -5,18 +5,21 @@
 namespace OpenTelemetry\SDK\Trace;
 
 use Exception;
+use GuzzleHttp\Client;
```
[the remark this answers is not in the document]
Comment (brettmc) [reply src/SDK/Trace/ExporterFactory.php:8]: I did have your in-flight work in mind when doing this. Right now, constructing most/all of the http-based exporters does require the creation of a client (in `::fromConnectionString`), and they all use Guzzle\Client, so I've done the same here.
Looking at it again, I think it is better if the client creation moves into the exporter itself, and can be ripped out again soon with http discovery. I will make that change.

File: tests/SDK/Unit/Trace/SamplerFactoryTest.php [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * @test
+     * @dataProvider samplerProvider
+     */
+    public function samplerFactory_createSamplerFromEnvironment(string $sampler, string $arg = null)
+    {
+        $this->setEnvironmentVariable('OTEL_TRACES_SAMPLER', $sampler);
+        $this->setEnvironmentVariable('OTEL_TRACES_SAMPLER_ARG', $arg);
+        $factory = new SamplerFactory();
+        $this->assertInstanceOf(SamplerInterface::class, $factory->fromEnvironment());
```
[the remark this answers is not in the document]
Comment (brettmc) [reply tests/SDK/Unit/Trace/SamplerFactoryTest.php]: that was a dummy assertion (and I found that the `@doesNotPerformAssertions` has a side-effect of not generating code coverage). It's prompted me to try harder though, so now it does some real assertions :)

File: tests/SDK/Unit/Trace/TracerProviderFactoryTest.php [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+ * @backupGlobals enabled
+ */
+class TracerProviderFactoryTest extends TestCase
+{
+    private $factory;
+    private $exporterFactory;
+    private $samplerFactory;
+    private $spanProcessorFactory;
+
+    public function setUp(): void
```
[the remark this answers is not in the document]
Comment (brettmc) [reply tests/SDK/Unit/Trace/TracerProviderFactoryTest.php]: they were used, but I've removed a pointless test and improved the remaining test, and I think it is ok now.

File: tests/SDK/Unit/Trace/TracerProviderFactoryTest.php [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        $this->assertInstanceOf(TracerProviderFactory::class, $factory);
+    }
+
+    /**
+     * @test
+     */
+    public function factory_createsExporter()
+    {
+        $tracer = $this->factory->create();
+        $this->assertInstanceOf(TracerProvider::class, $tracer);
```
[the remark this answers is not in the document]
Comment (brettmc) [reply tests/SDK/Unit/Trace/TracerProviderFactoryTest.php]: It was not a good test - there are not many public methods that I could use to check for success, but I settled on some assertions on the mock objects.

File: Makefile [hunk 16]
```diff
@@ -9,7 +9,7 @@ test:
 phan:
 	$(DC_RUN_PHP) env PHAN_DISABLE_XDEBUG_WARN=1 php ./vendor/bin/phan
 psalm:
-	$(DC_RUN_PHP) php ./vendor/bin/psalm --threads=1
+	$(DC_RUN_PHP) php ./vendor/bin/psalm --threads=1 --no-cache
```
[the remark this answers is not in the document]
Comment (brettmc) [reply Makefile:12]: I'm stumped by this - it should be really easy, and I wonder if it's my version of make (`GNU Make 4.2.1`), but nproc works locally, and in the project's image, but not via make.
I went back through history and found that threads was changed to 1 in #409 recently, so it might be safest to not play with it anyway!!

File: tests/SDK/Unit/Trace/SamplerFactoryTest.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (tidal) [reply tests/SDK/Unit/Trace/SamplerFactoryTest.php]: Yeah, I know the problem. That's why I spotted it. It's really tricky sometimes!

**Comment by brettmc** [comment]
@tidal - many thanks for the review and great feedback, I think I've addressed everything.

File: Makefile (hunk 16 above)
[the remark this answers is not in the document]
Comment (tidal) [reply Makefile:12]: Ok, no problem, no need to spend to much time on it for this PR. Thanks anyway!

File: src/SDK/Trace/TracerProviderFactory.php (hunk 4 above)
Comment (brettmc) [reply src/SDK/Trace/TracerProviderFactory.php:26]: My approach to type hinting is that it's documentation and safety (and it also helps IDEs), and if a method's return value is known then it should be type-hinted. I'm not sure why I wouldn't type-hint it, in theory we could create a second TracerProvider in the future and this factory would take care of determining which one to create...?

Comment (tidal) [reply src/SDK/Trace/TracerProviderFactory.php:26]: I was not wondering _why_ you use a type hint, but why it's TracerProviderInterface and not TracerProvider in this case. And since it's TracerProviderInterface, there could be a TracerProviderFactoryInterface as well. So this factory could implement the interface and we could use other factories using other methods in the future. (Not saying, that this has to be done in this PR though.)

Comment (brettmc) [reply src/SDK/Trace/TracerProviderFactory.php:26]: I see - I guess because I think it's better to code to an interface rather than a concrete implementation, and this factory could in theory create other trace providers, based on inputs that we have not invented yet!

Comment (tidal) [reply src/SDK/Trace/TracerProviderFactory.php:26]: Ok, got you.
Just for the record. I had also https://github.com/open-telemetry/opentelemetry-php/issues/301 in mind, while this is aiming at even another level above (SDK factory).

File: examples/ConfigurationFromEnvironment.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (tidal) [reply examples/ConfigurationFromEnvironment.php:20]: > (Or, did you just mean the example would be better with a try/catch to illustrate that it can throw exceptions?)

Yes, but if you want to work towards not throwing exceptions, it's not needed I guess.

**Comment by bobstrecansky** [comment]
There is one CI style issue @brettmc - otherwise I think we are looking good!

File: examples/ConfigurationFromEnvironment.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (tidal) [reply examples/ConfigurationFromEnvironment.php]: As I understood, _if_ it's an instrumentation using the SDK, the value should be the instrumenting/instrumented library/application name, but that's their resposibilty then.

**Comment by brettmc** [comment]
@bobstrecansky - all green and I'm happy with the PR now.

**Comment by bobstrecansky** [comment]
@tidal - I noticed you had some interweave here with PRs you're working on.  Are you good with this PR being merged?

**Comment by tidal** [comment]
@bobstrecansky 

> @tidal - Are you good with this PR being merged?

Yes, absolutely!

**Review by bobstrecansky — APPROVED** [review]
